core/src/main/java/io/parsingdata/metal/expression/value/NotAValue.java

```diff
@@ -0,0 +1,57 @@
+/*
+ * Copyright 2013-2016 Netherlands Forensic Institute
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.parsingdata.metal.expression.value;
+
+import java.math.BigInteger;
+import java.util.BitSet;
+import java.util.Objects;
+
+import io.parsingdata.metal.Util;
+import io.parsingdata.metal.data.Slice;
+import io.parsingdata.metal.encoding.Encoding;
+
+public final class NotAValue implements Value {
+
+    public static final Value NOT_A_VALUE = new NotAValue();
+    private static final UnsupportedOperationException UNSUPPORTED = new UnsupportedOperationException("NOT_A_VALUE does not support any Value operation.");
+
+    private NotAValue() {}
+
+    @Override public Slice getSlice() { throw UNSUPPORTED; }
+    @Override public Encoding getEncoding() { throw UNSUPPORTED; }
+    @Override public byte[] getValue() { throw UNSUPPORTED; }
+    @Override public BigInteger getLength() { throw UNSUPPORTED; }
+    @Override public BigInteger asNumeric() { throw UNSUPPORTED; }
+    @Override public String asString() { throw UNSUPPORTED; }
+    @Override public BitSet asBitSet() { throw UNSUPPORTED; }
+
+    @Override
+    public String toString() {
+        return "NOT_A_VALUE";
+    }
+
+    @Override
+    public boolean equals(final Object obj) {
+        return Util.notNullAndSameClass(this, obj);
```

Comment:
I think it could be a very intersting discussion wheter `NotAValue` should be equal to itself, which in this implementation it is. `NaN` compared to itself results into false. However, there are different opions about this if this is wrong or not. Still, many languages implements it this way.

We may consider to include a method `isValid()` or `isNotAValue()` in the `Value` and let `CoreValue` and `NotAValue` implement it. Then this method here could always return false and as such mimics the same behaviour as `NaN`. I'm not saying we should, but it is worth a thought. :)

Reply:
I've implemented that `NotAValue == NotAValue` results in `false`, but that is in the `Eq` operator, which is at the language-level of Metal, where these NaN-semantics are. At the Java implementation-level, `NOT_A_VALUE.equals(NOT_A_VALUE)` _must_ return `true` otherwise it breaks the defined semantics of all `Object`-implementations. For example, otherwise it would wreck maps that depend on the `equals(...)`-method to determine whether they already have a key.